// @axelar-network/axelarjs-sdk/dist/src/libs/TransactionRecoveryApi/helpers/contractEventHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findContractEvent = exports.getGasAmountFromTxReceipt = exports.getNativeGasAmountFromTxReceipt = exports.getLogIndexFromTxReceipt = exports.getDestinationChainFromTxReceipt = void 0;
const abi_1 = require("@ethersproject/abi");
const ethers_1 = require("ethers");
function getDestinationChainFromTxReceipt(receipt) {
    const signatureContractCallWithToken = ethers_1.ethers.utils.id("ContractCallWithToken(address,string,string,bytes32,bytes,string,uint256)");
    const signatureContractCall = ethers_1.ethers.utils.id("ContractCall(address,string,string,bytes32,bytes)");
    const event = findContractEvent(receipt, [signatureContractCall, signatureContractCallWithToken], new abi_1.Interface([
        "event ContractCallWithToken(address indexed sender, string destinationChain, string destinationContractAddress, bytes32 indexed payloadHash, bytes payload, string symbol, uint256 amount)",
        "event ContractCall(address indexed sender,string destinationChain,string destinationContractAddress,bytes32 indexed payloadHash,bytes payload)",
    ]));
    return event === null || event === void 0 ? void 0 : event.eventLog.args[1].toLowerCase();
}
exports.getDestinationChainFromTxReceipt = getDestinationChainFromTxReceipt;
function getLogIndexFromTxReceipt(receipt) {
    const signatureContractCallWithToken = ethers_1.ethers.utils.id("ContractCallWithToken(address,string,string,bytes32,bytes,string,uint256)");
    const signatureContractCall = ethers_1.ethers.utils.id("ContractCall(address,string,string,bytes32,bytes)");
    const event = findContractEvent(receipt, [signatureContractCall, signatureContractCallWithToken], new abi_1.Interface([
        "event ContractCallWithToken(address indexed sender, string destinationChain, string destinationContractAddress, bytes32 indexed payloadHash, bytes payload, string symbol, uint256 amount)",
        "event ContractCall(address indexed sender,string destinationChain,string destinationContractAddress,bytes32 indexed payloadHash,bytes payload)",
    ]));
    return event === null || event === void 0 ? void 0 : event.logIndex;
}
exports.getLogIndexFromTxReceipt = getLogIndexFromTxReceipt;
function getNativeGasAmountFromTxReceipt(receipt) {
    const signatureGasPaidContractCallWithToken = ethers_1.ethers.utils.id("NativeGasPaidForContractCallWithToken(address,string,string,bytes32,string,uint256,uint256,address)");
    const signatureGasPaidContractCall = ethers_1.ethers.utils.id("NativeGasPaidForContractCall(address,string,string,bytes32,uint256,address)");
    const event = findContractEvent(receipt, [signatureGasPaidContractCall, signatureGasPaidContractCallWithToken], new abi_1.Interface([
        "event NativeGasPaidForContractCall(address indexed sourceAddress,string destinationChain,string destinationAddress,bytes32 indexed payloadHash,uint256 gasFeeAmount,address refundAddress)",
        "event NativeGasPaidForContractCallWithToken(address indexed sourceAddress,string destinationChain,string destinationAddress,bytes32 indexed payloadHash,string symbol,uint256 amount,uint256 gasFeeAmount,address refundAddress)",
    ]));
    return event === null || event === void 0 ? void 0 : event.eventLog.args.slice(-2)[0].toString();
}
exports.getNativeGasAmountFromTxReceipt = getNativeGasAmountFromTxReceipt;
function getGasAmountFromTxReceipt(receipt) {
    const signatureGasPaidContractCallWithToken = ethers_1.ethers.utils.id("GasPaidForContractCallWithToken(address,string,string,bytes32,string,uint256,address,uint256,address)");
    const signatureGasPaidContractCall = ethers_1.ethers.utils.id("GasPaidForContractCall(address,string,string,bytes32,uint256,uint256,address)");
    const event = findContractEvent(receipt, [signatureGasPaidContractCall, signatureGasPaidContractCallWithToken], new abi_1.Interface([
        "event GasPaidForContractCallWithToken(address indexed sourceAddress,string destinationChain,string destinationAddress,bytes32 indexed payloadHash,string symbol,uint256 amount,address gasToken,uint256 gasFeeAmount,address refundAddress)",
        "event GasPaidForContractCall(address indexed sourceAddress,string destinationChain,string destinationAddress,bytes32 indexed payloadHash,address gasToken,uint256 gasFeeAmount,address refundAddress)",
    ]));
    return event === null || event === void 0 ? void 0 : event.eventLog.args.slice(-2)[0].toString();
}
exports.getGasAmountFromTxReceipt = getGasAmountFromTxReceipt;
function findContractEvent(receipt, eventSignatures, abiInterface) {
    for (const log of receipt.logs) {
        const eventIndex = eventSignatures.indexOf(log.topics[0]);
        if (eventIndex > -1) {
            const eventLog = abiInterface.parseLog(log);
            return {
                signature: eventSignatures[eventIndex],
                eventLog,
                logIndex: log.logIndex,
            };
        }
    }
}
exports.findContractEvent = findContractEvent;
